app/models/message.rb

```diff
@@ -9,11 +9,20 @@ class Message < ApplicationRecord
   scope :unread, -> { where(read: false) }
   scope :read, -> { where(read: true) }
 
+  after_create :update_conversation
+
   def mark_as_read!
     update!(read: true)
+    conversation.update!(has_unread_messages: false) if conversation.messages.unread.empty?
   end
 
   def unread?
     !read?
   end
+
+  private
+
+  def update_conversation
+    conversation.update(last_message_at: created_at, has_unread_messages: true)
```

Comment:
For data consistency, we probably want to ensure that if the conversation update fails for any reason, the message also doesn't get created. Adding a `!` into the conversation update would ensure the message doesn't get recorded unless the conversation gets updated:
 
```suggestion
    conversation.update!(last_message_at: created_at, has_unread_messages: true)
```